### PR TITLE
add trivia tiebreaker

### PR DESCRIPTION
**Pull request by josephmaxwellhyland**

### CODE REVIEW

File: back-end/db/friendsTriviaQuestions.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
+export default [
```
Comment (josephmaxwellhyland) [inline back-end/db/friendsTriviaQuestions.ts:1]: trivia about the TV show Friends (and one about Seinfeld) - this way, no player is the subject of the final question, which could cause issues

File: back-end/db/host.ts [hunk 2]
```diff
@@ -143,24 +144,28 @@ export default {
     return quizQuestions;
   },
 
+  questionsRemaining: function(game: IGame): boolean {
```
Comment (josephmaxwellhyland) [inline back-end/db/host.ts:147]: This checks if questions are remaining in the quiz

File: back-end/db/host.ts [hunk 3]
```diff
@@ -143,24 +144,28 @@ export default {
...
   nextQuestion: async function(gameId: number): Promise<boolean> {
     const currentGame: IGame | null = await this.getGameData(gameId);
     if (currentGame === null) {
       return false;
     }
 
-    const currentQuestionIndex = currentGame.currentQuestionIndex;
-    const nextQuestionIndex = currentQuestionIndex + 1;
+    if (this.questionsRemaining(currentGame)) {
+      await Game.updateOne({ id: gameId }, {
```
Comment (josephmaxwellhyland) [inline back-end/db/host.ts:161]: move the game onto the next question if there are more questions

File: back-end/db/host.ts [hunk 4]
```diff
@@ -306,4 +311,16 @@ export default {
       console.error(`Issue setting settings state: ${e}`);
     }
   },
+
+  addTiebreakerQuestion: async (gameId: number): Promise<void> => {
```
Comment (josephmaxwellhyland) [inline back-end/db/host.ts:315]: grab a random question from the friends questions

File: back-end/db/player.ts [hunk 5]
```diff
@@ -141,18 +141,24 @@ export default {
     }
   },
 
-  awardAllPlayersConsolationPoints: async function(guessingPlayers: IPlayer[], quizQIndex: number): Promise<any> {
+  getPlayersSortedByGuessSpeed: function(guessingPlayers: IPlayer[], quizQIndex: number): IPlayer[] {
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:144]: some refactoring here so this functionality can be re-used

File: back-end/db/player.ts [hunk 6]
```diff
@@ -168,18 +174,28 @@ export default {
...
       } else {    
         const newQuizGuesses = player.quizGuesses;
+        const correctGuess = guess === gameData.quizQuestions[gameData.currentQuestionIndex].correctAnswerIndex;
+        let scoreAdd = 0;
+        if (correctGuess) {
+          scoreAdd += 200;
+          const currentGuesses = await this.getPlayerGuessesForQuizQuestion(gameData.id, gameData.currentQuestionIndex);
+          const numGuesses = currentGuesses.filter(g => !!g).length;
+          if (numGuesses < 3) {
+            scoreAdd += (75 - numGuesses*25);
```
Comment (josephmaxwellhyland) [inline back-end/db/player.ts:196]: score bonus based on who answers first

File: back-end/handlers/hostHelpers.ts [hunk 7]
```diff
@@ -96,7 +88,7 @@ const hostStartQuizTimer = async (gameId: number, io: Server): Promise<void> =>
   nextQuestionTimer = setTimeout(hostPreAnswer, timePerQuestionMS, gameId, io);
 }
 
-const hostShowNextQuestion = async (gameId: number, io: Server): Promise<void> => {
+const hostNextQuestionOrLeaderboard = async (gameId: number, io: Server): Promise<void> => {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:91]: rename this to best reflect what it does - goes to a question OR the final leaderboard

File: back-end/handlers/hostHelpers.ts [hunk 8]
```diff
@@ -150,10 +142,35 @@ const hostPreAnswer = async (gameId: number, io: Server): Promise<void> => {
   setTimeout(hostShowAnswer, PRE_ANSWER_MS, gameId, io);
 }
 
+const handleTiebreakerAnswers = async (allPlayers: IPlayer[], quizQIndex, correctQIndex): Promise<void> => {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:145]: some ridiculous logic to handle all answers and ensure there is not another tie. basically, one of the tied players _has_ to get a bonus in one way or another.

File: back-end/handlers/hostHelpers.ts [hunk 9]
```diff
@@ -164,30 +181,35 @@ const hostShowAnswer = async (gameId: number, io: Server): Promise<void> => {
 
   const players = await playerDb.getPlayers(gameId);
   const subjectPlayerId = gameData.quizQuestions[currentQuestionIndex].playerId;
-  const guessingPlayers = players.filter(p => p.id !== subjectPlayerId);
+
+  const guessingPlayers = players.filter(p => p.id !== subjectPlayerId && p.quizGuesses[currentQuestionIndex]);
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:185]: small bug fix - have to make sure guessing players actually have a guess

File: back-end/handlers/hostHelpers.ts [hunk 10]
```diff
@@ -164,30 +181,35 @@ const hostShowAnswer = async (gameId: number, io: Server): Promise<void> => {
...
   const guesses = await playerDb.getPlayerGuessesForQuizQuestion(gameId, currentQuestionIndex);
 
   const correctGuess = gameData.quizQuestions[currentQuestionIndex].correctAnswerIndex;
   const totalGuesses = guesses.length - 1;
-  const numCorrect = guesses.filter(g => g &&g.guess === correctGuess).length;
+  const numCorrect = guesses.filter(g => g && g.guess === correctGuess).length;
 
-  if (numCorrect === 0) {
+  if (subjectPlayerId === "friends") {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:192]: if the question has this ID, it means the tiebreaker is happening

File: back-end/handlers/hostHelpers.ts [hunk 11]
```diff
@@ -164,30 +181,35 @@ const hostShowAnswer = async (gameId: number, io: Server): Promise<void> => {
...
   const subjectBonus = Math.floor(300 * (numCorrect / totalGuesses));
 
-  await Player.updateOne({
-    id: subjectPlayer.id
-  }, { 
-    $set: {
-      'score': subjectPlayer.score + subjectBonus
-    }
-  });
+  if (subjectPlayer) {
```
Comment (josephmaxwellhyland) [inline back-end/handlers/hostHelpers.ts:201]: only case when there's not a `subjectPlayer` is the tiebreaker - this handles that

File: front-end/src/host/HostShowAnswer.tsx [hunk 12]
```diff
@@ -49,19 +48,8 @@ export default function HostShowAnswer(props: IShowAnswerProps) {
...
-  }
-
   function onNext() {
-    if (currentQuizLength < quizLength) {
-      currentQuizLength++;
-      socket.emit("go-to-int-leaderboard", gameId);
-    } else {
-      socket.emit("next-question", gameId);
-    }
+    socket.emit("next-from-quiz-answer");
```
Comment (josephmaxwellhyland) [inline front-end/src/host/HostShowAnswer.tsx:52]: one message regardless, let the back-end figure out what to do

**Review by hylandtechoutreach — APPROVED** [review]
wow good work me

**Pull request merged**